### PR TITLE
JVM integration with InputStream and OutputStream

### PR DESCRIPTION
**Pull request by sandwwraith**

### CODE REVIEW

File: formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt
Comment (sandwwraith) [inline formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:63]: Probably we shouldn't expect eof here since there are use-cases for deserializing more than one object from stream (i.e. produce `Sequence<T>`)

File: formats/json/commonTest/src/kotlinx/serialization/json/BasicTypesSerializationTest.kt
Comment (qwwdfsad) [inline formats/json/commonTest/src/kotlinx/serialization/json/BasicTypesSerializationTest.kt:40]: It seems to be a typo from the previous version of the code, `useStreaming` should be used.
Also, it's worth renaming it to `jsonTestingMode` here

Comment (qwwdfsad) [reply formats/json/commonTest/src/kotlinx/serialization/json/BasicTypesSerializationTest.kt:40]: And in our test-specific `decode*` and `encode*` methods as well

File: formats/json/commonTest/src/kotlinx/serialization/json/JsonMapKeysTest.kt
Comment (qwwdfsad) [inline formats/json/commonTest/src/kotlinx/serialization/json/JsonMapKeysTest.kt:83]: parameter naming here as well

File: formats/json/commonTest/src/kotlinx/serialization/json/JsonTestBase.kt
Comment (qwwdfsad) [inline formats/json/commonTest/src/kotlinx/serialization/json/JsonTestBase.kt:28]: naming here and below

File: formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt
Comment (qwwdfsad) [inline formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:44]: A charset is quite a serious commitment in our public API.
Let's stick with UTF-8 only and document it as well?

File: formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt
Comment (qwwdfsad) [reply formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:63]: Right, this seems to be beyond the scope of our parsing.
In our string-related code, this assertion verified our own parsing logic is correct, but it's not our responsibility for streams.

I've checked out Jackson and it ignores content right after finished object as well

File: benchmark/src/jmh/kotlin/kotlinx/benchmarks/json/TwitterFeedStreamBenchmark.kt
Comment (qwwdfsad) [inline benchmark/src/jmh/kotlin/kotlinx/benchmarks/json/TwitterFeedStreamBenchmark.kt:35]: [nit] typically, for such benchmarks I prefer measurement from the stream that is already presented in the memory (i.e. BAIS/BAOS).

While it definitely not represent the "real"-ish workload, it does not skew our measurements and we can be sure that e.g. the benchmark score is not affected by the quality and/or load of the disk or file system,
In the end, the goal of this benchmark is to get precise (for this particular data) comparison with Jackson and we do not want our differences to be obscured or hidden by 3-rd party factors

File: benchmark/src/jmh/kotlin/kotlinx/benchmarks/json/TwitterFeedStreamBenchmark.kt
Comment (qwwdfsad) [inline benchmark/src/jmh/kotlin/kotlinx/benchmarks/json/TwitterFeedStreamBenchmark.kt:22]: I see that a lot of `decode*` benchmarks dropped their score by ~15%.

It seems to be a serious hit and it's definitely worth pinpointing it before merging this PR, so we are sure we cannot avoid it (or can and then will do)

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt
Comment (qwwdfsad) [inline formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt:79]: excessive indent

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt
Comment (qwwdfsad) [inline formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt:28]:
```suggestion
internal actual open class JsonStringBuilder(@JvmField protected var array: CharArray) {
```

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonToWriterStringBuilder.kt
Comment (qwwdfsad) [inline formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonToWriterStringBuilder.kt:10]: It seems to be pretty huge allocation (e.g. it always misses TLAB). Could you please ensure it doesn't dominate small objects serialization?

If so, it's worth either reducing its size or pool a few instances of this

**Review by qwwdfsad — CHANGES_REQUESTED** [review]
I haven't got a deep look into decoding yet, but it seems that there is already enough actionable points here

**Comment by chris-hatton** [comment]
Question regarding the behaviour of this change.  Will `decodeFromStream` behave such that I can call it serially against the same stream, each time consuming only as much of the stream as is necessary to form the complete object?
e.g. given the stream `{"someKey":"someValue1"}{"someKey":"someValue2"}`
I can call `decodeFromStream()` twice to get both objects?
This is the characteristic I am looking for, to be able to read a `Flow<T>` from a long-lived HTTP response stream.
Thanks for your efforts @sandwwraith 🙏 This is a hotly awaited improvement.

File: formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt
Comment (JakeWharton) [reply formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:63]: Moshi will leave your stream intact allowing you to do things like JSON-RPC which has newline-delimited objects. Granted, we also require that you pass in a buffered stream so that we can read only what's necessary and leave everything else in the buffer you then have access to.

**Comment by BenWoodworth** [comment]
I'm curious, was using [`Reader`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/Reader.html)/[`Writer`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/Writer.html)s considered (instead of `Stream`s)? That would've been my first thought for JSON/String formats, and avoids [dealing with character encodings](https://github.com/Kotlin/kotlinx.serialization/pull/1569/files/8678e127d653b334ff5288fd6b617cbbd02c633c#r659932844).

File: formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt
Comment (chris-hatton) [reply formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:63]: I have a use case for implementing a `Flow<T>` from a long-lived HTTP response stream.  To have the lexer expect the end of the stream here seems wrong - why impose this opinion on the stream?  The lexers interest in the stream should begin and end with the fully-formed JSON content IMO

Comment (pdvrieze) [reply formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:63]: having thought about the streaming case I would say that a flow of objects would be a valid case to handle. Probably flow is the right type - both for encoding and decoding. There are some interesting interactions with coroutines there though and flows should probably have their own entrypoint.

Comment (pdvrieze) [reply formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:63]: As a separate matter, these are provided as extension functions. It would be interesting to have an interface like `StringFormat` or `BinaryFormat` (or a combination interface) that provides the consistency between formats. There is also a case to support the `Reader` and `Writer` interfaces next to/instead of InputStream/OutputStream (that would also avoid bringing in charset explicitly while still allowing non-utf8 to be used).

**Comment by sandwwraith** [comment]
@BenWoodworth `Reader/Writer` is used internally. API provides methods with `Input/OutputStream`s because it's more versatile and allows to implement charset-specific parsers in the future

@chris-hatton Yes, I think we can do it — I'll add a test for it

File: formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt
Comment (sandwwraith) [reply formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:44]: I don't see many problems in it — Charset isn't going anywhere from `java.nio`, AFAIK. We still can implement charset-specific decoders if we need to.

Of course, we also can add overload with charset later, if there would be demand for it, but we'll end up with two almost identical overloads instead of a function with default parameter

Comment (qwwdfsad) [reply formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:44]: It will obstruct us from properly optimizing UTF-8 decoding because we will be forced to maintain the rest of the charsets (think about suffix tree), apart from that charsets are quite slow on their own.

Moreover, most of the modern applications are using UTF-8 anyway and it's unlikely we'll stumble across such demand. So I suggest avoiding introducing charsets into our API and rely on UTF-8 only, waiting for a proper demand.

**Review by qwwdfsad — CHANGES_REQUESTED** [review]
I've dug through the profile and it seems that JVM cannot optimize string access across `CharSequence` interface properly.

I've tried to tweak it here and there, but it's quite hard to ensure all the invariants with the existing limitations.

I'd suggest you to do the following:

Get the base JsonLexer with the only state -- `currentPosition` and utility functions for slow-paths: `skipElement`, various `fail` functions, maybe boolean/numbers consumption (char sequence is an input parameter of a function then). Everything else is copy-pasted between streaming and string implementations.

At this moment, the performance model is quite clear and expected degradation should be insignificant (educated guess -- 2-4%).
Then you can start commonizing (handling via `CharSequence` interface in the base json lexer) the parts of parsing where the compiler is smart enough to optimize everything away.

I expect that the biggest offenders (things you cannot commonize) will be just a few functions that were written in a compact and polished manner -- `skipWhitespaces`, `tryConsumeComma`, `canConsumeValue` and `peekNextToken`. Everything else will probably be working as expected

File: formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt
Comment (pdvrieze) [reply formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:44]: @qwwdfsad I agree on the character set thing. Especially in the context of taking a "writer". The api to handle the charsets is already there. The only counterpoint is the speed of parsing streams vs readers.

Comment (qwwdfsad) [reply formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:44]: Good to know, thanks!

Leonid is on the vacation right now, so the final decision is not yet made, but we've discussed this and other cases on our design meeting and seems that we have a consensus regarding encodings -- UTF-8 dominates all along, and everyone is mostly fine to be UTF-8 first. Even Java changes towards that: [JEP-400](https://openjdk.java.net/jeps/400). Moshi is also a good example of that.

So it's _likely_ that all our new API will be using UTF-8 as well and other encoding will be put aside and be supported on a best-effort basis

File: formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt
Comment (qwwdfsad) [inline formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt:12]: Redundant dot

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonToWriterStringBuilder.kt
Comment (qwwdfsad) [inline formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonToWriterStringBuilder.kt:31]: [nit] it's just `flush` :)

**Review by qwwdfsad — APPROVED** [review]
Good to go 🚀 

Please don't forget to file issues for future improvements -- UTF-8 parsing and multishot streams

**Comment by sandwwraith** [comment]
https://github.com/Kotlin/kotlinx.serialization/issues/1662

**Comment by slavonnet** [comment]
String. Serializer work ok on big JSon. Stream   get  random.eof exception  I sinppe replace Strung Encoder t to  Stream Encoder  in converter factory . Small size json is ok. Buffer rewrite by gzip wtite all

**Pull request merged**
